File: src/main/java/io/github/phantamanta44/libnine/util/nbt/ChainingTagCompound.java
package io.github.phantamanta44.libnine.util.nbt;

import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTBase;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraftforge.fluids.FluidStack;
import scala.actors.threadpool.Arrays;

import java.util.List;
import java.util.function.Consumer;

public class ChainingTagCompound extends NBTTagCompound {

    public ChainingTagCompound() {
        // NO-OP
    }

    public ChainingTagCompound(NBTTagCompound base) {
        for (String key : base.getKeySet()) setTag(key, base.getTag(key));
    }

    public ChainingTagCompound withTag(String key, NBTTagCompound tag) {
        super.setTag(key, tag);
        return this;
    }

    public ChainingTagCompound withInt(String key, int value) {
        super.setInteger(key, value);
        return this;
    }

    public ChainingTagCompound withByte(String key, int value) {
        super.setByte(key, (byte)value);
        return this;
    }

    public ChainingTagCompound withFloat(String key, float value) {
        super.setFloat(key, value);
        return this;
    }

    public ChainingTagCompound withDouble(String key, double value) {
        super.setDouble(key, value);
        return this;
    }

    public ChainingTagCompound withBool(String key, boolean value) {
        super.setBoolean(key, value);
        return this;
    }

    public ChainingTagCompound withStr(String key, String value) {
        super.setString(key, value);
        return this;
    }

    @SuppressWarnings("unchecked")
    public <T extends NBTBase> ChainingTagCompound withList(String key, T... tags) {
        return withList(key, Arrays.asList(tags));
    }

    public <T extends NBTBase> ChainingTagCompound withList(String key, List<T> tags) {
        NBTTagList list = new NBTTagList();
        for (T tag : tags) {
            list.appendTag(tag);
        }
        return withList(key, list);
    }

    public ChainingTagCompound withList(String key, NBTTagList list) {
        setTag(key, list);
        return this;
    }

    public ChainingTagCompound withItemStack(String key, ItemStack value) {
        return withSerializable(key, value::writeToNBT);
    }

    public ChainingTagCompound withFluidStack(String key, FluidStack value) {
        return withSerializable(key, value::writeToNBT);
    }

    public ChainingTagCompound withSerializable(String key, Consumer<NBTTagCompound> writer) {
        super.setTag(key, new NBTTagCompound());
        writer.accept(super.getCompoundTag(key));
        return this;
    }

}

File: src/main/kotlin/xyz/phanta/libnine/util/data/nbt/ChainingTagCompound.kt
package xyz.phanta.libnine.util.data.nbt

import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundNBT
import net.minecraft.nbt.INBT
import net.minecraft.nbt.ListNBT
import net.minecraftforge.fluids.FluidStack

class ChainingTagCompound : CompoundNBT() {

    fun withTag(key: String, tag: CompoundNBT): ChainingTagCompound = also { super.put(key, tag) }

    fun withInt(key: String, value: Int): ChainingTagCompound = also { super.putInt(key, value) }

    fun withByte(key: String, value: Int): ChainingTagCompound = also { super.putByte(key, value.toByte()) }

    fun withFloat(key: String, value: Float): ChainingTagCompound = also { super.putFloat(key, value) }

    fun withDouble(key: String, value: Double): ChainingTagCompound = also { super.putDouble(key, value) }

    fun withBool(key: String, value: Boolean): ChainingTagCompound = also { super.putBoolean(key, value) }

    fun withStr(key: String, value: String): ChainingTagCompound = also { super.putString(key, value) }

    fun <T : INBT> withList(key: String, vararg tags: T): CompoundNBT = also {
        super.put(key, ListNBT().also { list ->
            tags.forEach { list.add(it) }
        })
    }

    fun withItemStack(key: String, value: ItemStack): ChainingTagCompound = withSerializable(key) { value.write(it) }

    fun withFluidStack(key: String, value: FluidStack): ChainingTagCompound = TODO("forge doesn't have a fluid impl yet")

    fun withSerializable(key: String, writer: (CompoundNBT) -> Unit): ChainingTagCompound = also {
        super.put(key, CompoundNBT().also { writer(it) })
    }

}
